File: examples/database_example.cpp
#include "database_manager.h"
#include <iostream>

/**
 * @brief Example demonstrating database integration with Conflict-Free Scheduling
 */
int main() {
    std::cout << "=== Database Integration Example ===" << std::endl;
    
    // Initialize database manager
    DatabaseManager dbManager("data/scheduling.db");
    
    if (!dbManager.initialize()) {
        std::cerr << "Failed to initialize database: " << dbManager.getLastError() << std::endl;
        return 1;
    }
    
    std::cout << "✓ Database initialized successfully" << std::endl;
    
    // Reset and load sample data
    if (!dbManager.resetDatabase() || !dbManager.loadSampleData()) {
        std::cerr << "Failed to load sample data: " << dbManager.getLastError() << std::endl;
        return 1;
    }
    
    std::cout << "✓ Sample data loaded" << std::endl;
    
    // Display database statistics
    auto stats = dbManager.getScheduleStatistics();
    std::cout << "\n=== Database Statistics ===" << std::endl;
    std::cout << "Total Courses: " << stats.total_courses << std::endl;
    std::cout << "Total Rooms: " << stats.total_rooms << std::endl;
    std::cout << "Total Time Slots: " << stats.total_time_slots << std::endl;
    std::cout << "Scheduled Courses: " << stats.scheduled_courses << std::endl;
    std::cout << "Unscheduled Courses: " << stats.unscheduled_courses << std::endl;
    std::cout << "Unresolved Conflicts: " << stats.conflicts_detected << std::endl;
    std::cout << "Resolved Conflicts: " << stats.conflicts_resolved << std::endl;
    std::cout << "Room Utilization: " << stats.room_utilization_percentage << "%" << std::endl;
    
    // Get all courses from database
    auto courses = dbManager.getAllCourses();
    std::cout << "\n=== Course List ===" << std::endl;
    for (const auto& course : courses) {
        std::cout << "Course ID: " << course.id 
                  << ", Name: " << course.name 
                  << ", Duration: " << course.duration 
                  << ", Priority: " << course.priority << std::endl;
    }
    
    // Example: Insert a new course
    Activity newCourse;
    newCourse.name = "CSE500-Advanced Algorithms";
    newCourse.duration = 3;
    newCourse.priority = 1;
    newCourse.start_time = 0;
    newCourse.end_time = 3;
    
    if (dbManager.insertCourse(newCourse)) {
        std::cout << "\n✓ Successfully inserted new course: " << newCourse.name << std::endl;
    } else {
        std::cerr << "\n✗ Failed to insert course: " << dbManager.getLastError() << std::endl;
    }
    
    // Example: Conflict logging
    DatabaseManager::ConflictLog conflict;
    conflict.conflict_type = "ROOM_CAPACITY";
    conflict.description = "Course enrollment exceeds room capacity";
    conflict.course_id_1 = 1;
    conflict.course_id_2 = 0;
    conflict.room_id = 1;
    conflict.slot_id = 1;
    conflict.resolved = false;
    
    if (dbManager.logConflict(conflict)) {
        std::cout << "\n✓ Conflict logged successfully" << std::endl;
    }
    
    // Get unresolved conflicts
    auto conflicts = dbManager.getUnresolvedConflicts();
    std::cout << "\n=== Unresolved Conflicts ===" << std::endl;
    for (const auto& c : conflicts) {
        std::cout << "Conflict ID: " << c.conflict_id 
                  << ", Type: " << c.conflict_type
                  << ", Description: " << c.description << std::endl;
    }
    
    std::cout << "\n=== Database Example Complete ===" << std::endl;
    return 0;
}

File: examples/database_example.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Database Integration Example for Conflict-Free Scheduling System

This example demonstrates how to use the database functionality
of the scheduling system.
"""

import sys
import os

# Add src to path for imports
sys.path.append(os.path.join(os.path.dirname(__file__), '..', 'src'))

try:
    from database.database_manager import DatabaseManager, Activity
    from scheduler import ConflictFreeScheduler
except ImportError as e:
    print(f"❌ Import error: {e}")
    print("Make sure to install dependencies: pip install sqlalchemy pymysql")
    sys.exit(1)


def main():
    """Main database example function"""
    print("=== Database Integration Example ===")
    
    # Initialize database manager
    print("🔄 Initializing database...")
    db_manager = DatabaseManager(
        host="localhost",
        database="conflict_free_scheduling",
        user="root",
        password=""  # Update with your MySQL password
    )
    
    # Initialize database with sample data
    print("📊 Setting up database with sample data...")
    if not db_manager.initialize_with_sample_data():
        print("❌ Failed to initialize database")
        print("Make sure MySQL is running and you have the correct credentials")
        return 1
    
    print("✅ Database initialized successfully")
    
    # Get database statistics
    print("\n=== Database Statistics ===")
    try:
        session = db_manager.Session()
        
        # Count records
        from database.database_manager import Batch, Teacher, Classroom, Course
        
        batch_count = session.query(Batch).count()
        teacher_count = session.query(Teacher).count()
        room_count = session.query(Classroom).count()
        course_count = session.query(Course).count()
        
        print(f"Total Batches: {batch_count}")
        print(f"Total Teachers: {teacher_count}")
        print(f"Total Rooms: {room_count}")
        print(f"Total Courses: {course_count}")
        
        session.close()
        
    except Exception as e:
        print(f"❌ Error getting statistics: {e}")
        return 1
    
    # Get all courses as activities
    print("\n=== Course List ===")
    activities = db_manager.get_all_courses()
    
    if activities:
        print(f"Found {len(activities)} courses:")
        for i, activity in enumerate(activities[:5]):  # Show first 5
            print(f"  {i+1}. {activity.name} (Weight: {activity.weight}, Duration: {activity.end - activity.start})")
        
        if len(activities) > 5:
            print(f"  ... and {len(activities) - 5} more courses")
    else:
        print("No courses found in database")
        return 1
    
    # Test scheduling with database data
    print("\n=== Scheduling with Database Data ===")
    scheduler = ConflictFreeScheduler()
    
    # Run graph coloring algorithm
    print("🔄 Running Graph Coloring algorithm...")
    scheduled = scheduler.graph_coloring_schedule(activities)
    
    if scheduled:
        total_weight = scheduler.calculate_total_weight(scheduled)
        efficiency = (len(scheduled) / len(activities)) * 100
        
        print(f"✅ Scheduled {len(scheduled)}/{len(activities)} courses")
        print(f"📊 Total credit hours: {total_weight:.1f}")
        print(f"📈 Efficiency: {efficiency:.1f}%")
        
        # Show first few scheduled courses
        print("\nScheduled courses:")
        for i, activity in enumerate(scheduled[:3]):
            print(f"  {i+1}. {activity.name} - Time: {activity.start}-{activity.end}")
        
        if len(scheduled) > 3:
            print(f"  ... and {len(scheduled) - 3} more")
    else:
        print("❌ No courses could be scheduled")
    
    # Test batch-specific scheduling
    print("\n=== Batch-Specific Scheduling ===")
    batch_activities = db_manager.get_courses_by_batch("BCSE24")
    
    if batch_activities:
        print(f"Found {len(batch_activities)} courses for BCSE24")
        batch_scheduled = scheduler.graph_coloring_schedule(batch_activities)
        batch_weight = scheduler.calculate_total_weight(batch_scheduled)
        
        print(f"✅ Scheduled {len(batch_scheduled)} BCSE24 courses")
        print(f"📊 Total credit hours: {batch_weight:.1f}")
    else:
        print("No courses found for BCSE24")
    
    # Close database connection
    db_manager.close()
    print("\n✅ Database example completed successfully!")
    return 0


if __name__ == "__main__":
    sys.exit(main())
